distribution/lib/Standard/Table/0.0.0-dev/src/Io/Excel.enso

```diff
@@ -0,0 +1,157 @@
+
+        all_valid <| Excel_Range (Java_Range.new sheet col_index row col_index row)
+
+    ## Create a Range for a range of cells.
+    for_range : Text -> (Text|Integer) -> Integer -> (Text|Integer) -> Integer -> Excel_Range
+    for_range sheet left top right bottom =
+        left_index = Excel_Range.column_index left
+        right_index = Excel_Range.column_index right
+
+        left_valid = here.validate (Excel_Range.is_valid_column left_index) ("Invalid left column for Excel: " + left.to_text + ".") _
+        right_valid = here.validate (Excel_Range.is_valid_column right_index) ("Invalid right column for Excel: " + right.to_text + ".") (left_valid _)
+        top_valid = here.validate (Excel_Range.is_valid_row top) ("Invalid top row for Excel: " + top.to_text + ".") (right_valid _)
+        all_valid = here.validate (Excel_Range.is_valid_row bottom) ("Invalid bottom row for Excel: " + bottom.to_text + ".") (top_valid _)
+
+        all_valid <| Excel_Range (Java_Range.new sheet left_index top right_index bottom)
+
+    ## Create an Excel_Range for a set of columns.
+    for_columns : Text -> (Text|Integer) -> (Text|Integer) -> Excel_Range
+    for_columns sheet left (right=left) =
+        left_index = Excel_Range.column_index left
+        right_index = Excel_Range.column_index right
+
+        left_valid = here.validate (Excel_Range.is_valid_column left_index) ("Invalid left column for Excel: " + left.to_text + ".") _
+        all_valid = here.validate (Excel_Range.is_valid_column right_index) ("Invalid right column for Excel: " + right.to_text + ".") (left_valid _)
+
+        all_valid <| Excel_Range (Java_Range.new sheet left_index 0 right_index 0)
+
+    ## Create an Excel_Range for a set of rows.
+    for_rows : Text -> Integer -> Integer -> Excel_Range
+    for_rows sheet top (bottom=top) =
+        top_valid = here.validate (Excel_Range.is_valid_row top) ("Invalid top row for Excel: " + top.to_text + ".") _
+        all_valid = here.validate (Excel_Range.is_valid_row bottom) ("Invalid bottom row for Excel: " + bottom.to_text + ".") (top_valid _)
+
+        all_valid <| Excel_Range (Java_Range.new sheet 0 top 0 bottom)
+
+
+## PRIVATE
+   Wrapper for validation
+validate : Boolean -> Text -> Any
+validate ~validation ~error_message ~wrapped =
+    if validation then wrapped else Error.throw (Illegal_Argument_Error error_message)
+
+read_excel : File -> Excel_Section -> Problem_Behavior -> Boolean -> (Table | Vector)
+read_excel file section _ xls_format=False =
+    reader stream = case section of
+        Sheet_Names -> Vector.Vector (Reader.readSheetNames stream xls_format)
+        Range_Names -> Vector.Vector (Reader.readRangeNames stream xls_format)
+        Sheet sheet skip_rows row_limit ->
+            Table.Table <|
+                if sheet.is_an Integer then (Reader.readSheetByIndex stream sheet skip_rows row_limit xls_format) else
+                    Reader.readSheetByName stream sheet skip_rows row_limit xls_format
```

Comment:
```suggestion
            Table.Table <| case sheet of
                Integer -> Reader.readSheetByIndex stream sheet skip_rows row_limit xls_format
                Text -> Reader.readSheetByName stream sheet skip_rows row_limit xls_format
```

Why use `is_an` (which currently has a surprisingly complex implementation) if pattern matching can be used? IMHO it looks even a bit clearer, and also gives us slightly more type safety (we can check that the alternative branch is `Text` and if something else arrives a pattern match error will be raised here)

and same below

Reply:
nested cases have been tetchy so generally have avoided.

Will amend and see if works.

Reply:
> nested cases have been tetchy so generally have avoided.
> 
> Will amend and see if works.

oh...

But wasn't this only the case for nested patterns? i.e. stuff like
```
case x of
    Pair True False -> ...
```

I think a lot of code nests independent `case`s without issues. But I may be mistaken here.